### PR TITLE
Add support for array variable type validation

### PR DESCRIPTION
**Pull request by rw-access**
## Issues
Alternative approach to #68

## Details
Basically I just updated the parser so that the behavior for loop variables implicitly is relative to an array in the directly previous argument. I don't think there are any exceptions, and if there are they can be dealt with as they occur. This approach is a bit hacky but simple enough of a change to maintain.

I started going down a rabbit hole of rewriting most of the type system, but that scope was hard to cut so I moved on to this more simplistic approach where variable behavior is implicitly defined.

### CODE REVIEW

**Review by brokensound77 — APPROVED** [review]
Tested fine. Bump version and update change log and LGTM

**Comment by Mikaayenson** [comment]
👋 @rw-access Thanks for adding this PR! Can you add a test case using the $variable syntax with the arraysearch function?

File: eql/parser.py [hunk 1]
```diff
@@ -442,7 +442,16 @@ def _update_field_info(self, node_info, optional_syntax=False):
                     return NodeInfo(comparison, TypeHint.Boolean, nullable=not self._strict_fields, source=node_info)
 
         else:
-            type_hint = self._var_types[field.base]
+            # Extract out types from variables and perform validation on the nested schema
+            schema_hint = self._var_types[field.base]
```
Comment (Mikaayenson) [inline eql/parser.py:446]: In which case is this logic hit re arraysearch use?

Comment (rw-access) [reply eql/parser.py:446]:
```
arraySearch(someArray, $myvar, $myvar.foo == "bar")
                               ^^^^^^
```

File: eql/parser.py [hunk 2]
```diff
@@ -880,11 +889,14 @@ def function_call(self, node, prev_node=None, prev_arg=None):
...
             # and we need to pull out all of the variables
+            var_type = None
+            var_schema = None
+
             for idx, arg_source in enumerate(nodes):
                 if idx in variables:
                     if arg_source.data == "base_field":
                         variable_name = self.visit(arg_source["name"])
-                        self._add_variable(variable_name)
+                        self._add_variable(variable_name, var_type or TypeHint.Unknown, var_schema)
```
Comment (Mikaayenson) [inline eql/parser.py:899]: Do we have to add an special case for `arg_source.data  == 'varpath'`?

Comment (rw-access) [reply eql/parser.py:899]: What do you mean? Can you give an example?

Comment (Mikaayenson) [reply eql/parser.py:899]: Screenshot 2023-07-21 at 5 01 36 PM

Right now the logic is erroring when arg_source.data is `varpath`. Is that expected or should we add another else if statement for the `$var`.

**Comment by Mikaayenson** [comment]
## Update July 24

Should be g2g now. Catches bad nested fields and nested field types used in the arraySearch function.

**Review by brokensound77 — APPROVED** [review]
LGTM, thanks y'all

**Pull request merged**